### PR TITLE
Only remove attributes with empty values when updating user profile

### LINKED ISSUE #27797: User profile fields cannot be set empty once they have a non-empty value (in Login Theme)
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

user-profile

### Describe the bug

I have observed that it's currently not possible to set a non-empty field to empty by means of the "Update profile" action (and I guess the same applies to the other Login Theme actions).
To be more precise, the UI allows to set an empty value (when the field is not required), but the empty value is not stored at server side. In both Admin and Account UI the old (non-empty) value is still displayed.

### Version

23.0.7 AND current main branch (commit 1c0c56e8055ef227d3aa72102a2acb10eb306d16)

### Regression

- [ ] The issue is a regression

### Expected behavior

- When a field is not required, it should be possible to set the field to empty, and the empty value SHOULD BE persisted.

### Actual behavior

When a field is not required, it is possible to set the field to empty, but the empty value is NOT persisted.

### How to Reproduce?

- In the Admin UI (or via Admin API): 
  - Add a simple optional attribute to a realm's user profile - let's call it `test`.
  - Create a user at the realm, set the required action "Update Profile" for the user, and initialize the `test` attribute value with "anything".
- In the Login Theme / Account UI:
  - Login at the Account UI with the user.
  - The "Update Profile" screen should be displayed with `test` attribute value "anything".
  - Clear the `test` attribute field and submit the form.
  - The Account UI should now be displayed. The `test` attribute value displayed there is still "anything". The same applies when displaying the user in the Admin UI. 

### Anything else?

Note that I've tested this with version 23.0.7 and current main branch. Probably this also occurs with version 24.x.

**UPDATE**: Clearing values with the Account UI and Admin UI works fine. The issue seems to be restricted to the Login Theme.

### PR DESCRIPTION
**Pull request by pedroigor**
Closes #27797

### CODE REVIEW

**Review by sguilhen — APPROVED** [review]

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/user/profile/UserProfileTest.java [hunk 1]
```diff
@@ -235,6 +235,47 @@ private static void testCustomAttributeInAnyContext(KeycloakSession session) {
...
+        attributes.put("address", "");
+        profile = provider.create(UserProfileContext.UPDATE_PROFILE, attributes, user);
+        profile.update(false);
+
+        assertNull(user.getFirstAttribute("address"));
+
+        attributes.put("address", "bar");
+        profile = provider.create(UserProfileContext.UPDATE_PROFILE, attributes, user);
+        profile.update();
+
```
Comment (danielFesenmeyer) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/user/profile/UserProfileTest.java:270]: What about adding a "pre-condition check" for the next assertion here:

`assertNotNull(user.getFirstAttribute("address"));`

Just to make sure that "bar" has actually been stored.

File: model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java [hunk 2]
```diff
@@ -196,14 +196,22 @@ private void persistAttributeValue(String name, String value) {
...
+        List<UserAttributeEntity> customAttributesToRemove = new ArrayList<>();
         for (UserAttributeEntity attr : user.getAttributes()) {
             if (attr.getName().equals(name)) {
-                toRemove.add(attr);
+                customAttributesToRemove.add(attr);
             }
         }
 
-        if (toRemove.isEmpty()) {
+        if (customAttributesToRemove.isEmpty()) {
```
Comment (danielFesenmeyer) [inline model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java:206]: This branch makes sense from my POV, but I don't get how it relates to issue #27797.
Is this an additional improvement?

**Review by danielFesenmeyer — COMMENTED** [review]
Hi @pedroigor, thanks for your PR :smiley:
LGTM :thumbsup:
Just added two small comments.

File: model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java (hunk 2 above)
Comment (pedroigor) [reply model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java:206]: Yeah, it is. We were missing removing the "root" attributes from the user but only custom attributes. Every other method related to managing attributes do something similar. So I just doing the same here for the sake of consistency.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/user/profile/UserProfileTest.java (hunk 1 above)
Comment (pedroigor) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/user/profile/UserProfileTest.java:270]: I'll add. We already have similar assertions in other test methods (and in the lines right after)  but we can add that.

**Review by rmartinc — APPROVED** [review]
LGTM!

**Comment by danielFesenmeyer** [comment]
@pedroigor Thanks a lot for this fix!

**Review by mposolda — APPROVED** [review]

**Pull request merged**